File: backend/Minimarket.PrintWorker/Services/TicketTextRenderer.cs
using System.Globalization;
using System.Text;
using Minimarket.PrintWorker.DTOs;
using Minimarket.PrintWorker.Options;

namespace Minimarket.PrintWorker.Services;

public class TicketTextRenderer
{
    public string Build(TicketPrintPayload payload, PrintingOptions options)
    {
        var width = Math.Max(32, options.LineWidth);

        // Los datos vienen del snapshot (igual que la vista previa). Si un ticket viejo no los trae,
        // se usan como respaldo las opciones locales del worker.
        var businessName = Fallback(payload.BusinessName, options.BusinessName);
        var legalName = Fallback(payload.LegalName, options.LegalName);
        var taxId = Fallback(payload.TaxId, options.TaxId);
        var addressLine = Fallback(payload.AddressLine, options.AddressLine);
        var phone = Fallback(payload.Phone, options.Phone);
        var tagline = payload.Tagline?.Trim() ?? string.Empty;
        var documentTitle = Fallback(payload.DocumentTitle, "Ticket de venta");
        var customerLabel = Fallback(payload.CustomerLabel, options.CustomerLabel);
        var footerLine1 = Fallback(payload.FooterLine1, options.FooterLine1);
        var footerLine2 = Fallback(payload.FooterLine2, options.FooterLine2);

        // Totales del snapshot; si un ticket viejo no trae el desglose, se deriva del total (IGV 18% incluido).
        var total = payload.Total;
        var subTotal = payload.SubTotal > 0 ? payload.SubTotal : decimal.Round(total / 1.18m, 2, MidpointRounding.AwayFromZero);
        var igv = payload.Igv > 0 ? payload.Igv : decimal.Round(total - subTotal, 2, MidpointRounding.AwayFromZero);

        var lines = new List<string> { Center(businessName, width) };
        if (!string.IsNullOrWhiteSpace(tagline))
        {
            lines.Add(Center(tagline, width));
        }

        lines.Add(new string('-', width));
        lines.Add(Center(legalName, width));
        lines.Add(Center($"RUC: {taxId}", width));
        lines.Add(Center(addressLine, width));
        lines.Add(Center($"Telefono: {phone}", width));
        lines.Add(new string('-', width));
        lines.Add(Center(documentTitle.ToUpperInvariant(), width));
        lines.Add(Center($"#{payload.SaleId} - {payload.SaleDate:dd/MM/yyyy HH:mm}", width));
        lines.Add(new string('-', width));
        lines.Add($"Fecha: {payload.SaleDate:dd/MM/yyyy HH:mm}");
        lines.Add($"Cajero: {payload.CashierName}");
        lines.Add($"Cliente: {customerLabel}");
        lines.Add($"Pago: {payload.PaymentMethod}");
        lines.Add(new string('-', width));
        lines.Add(TicketHeader(width));
        lines.Add(new string('-', width));

        foreach (var item in payload.Items)
        {
            foreach (var wrapped in Wrap(item.ProductName, width - 2))
            {
                lines.Add(wrapped);
            }

            var detail = $"{item.Quantity} x {FormatMoney(item.UnitPrice)}";
            lines.Add(PadRight(detail, Math.Max(0, width - FormatMoney(item.Subtotal).Length)) + FormatMoney(item.Subtotal));
        }

        lines.Add(new string('-', width));
        lines.Add(Row("Items", payload.Items.Count.ToString(CultureInfo.InvariantCulture), width));
        var totalUnits = payload.Items.Sum(item => item.Quantity);
        lines.Add(Row("Unidades", totalUnits.ToString(CultureInfo.InvariantCulture), width));
        lines.Add(Row("Subtotal", FormatMoney(subTotal), width));
        lines.Add(Row("IGV (18%)", FormatMoney(igv), width));

        if (!string.IsNullOrWhiteSpace(payload.Notes))
        {
            lines.Add(new string('-', width));
            lines.Add("Notas:");
            lines.AddRange(Wrap(payload.Notes, width));
        }

        lines.Add(new string('-', width));
        lines.Add(Row("TOTAL", FormatMoney(total), width));

        var builder = new StringBuilder();
        foreach (var line in lines)
        {
            builder.AppendLine(line);
        }

        builder.AppendLine();
        builder.AppendLine(Center(footerLine1, width));
        builder.AppendLine(Center(footerLine2, width));
        builder.AppendLine();
        builder.AppendLine();
        return builder.ToString();
    }

    private static string Fallback(string? value, string fallback) =>
        string.IsNullOrWhiteSpace(value) ? fallback : value.Trim();

    private static string Row(string label, string value, int width) =>
        PadRight(label, Math.Max(0, width - value.Length)) + value;

    private static string FormatMoney(decimal value) =>
        $"S/ {value:0.00}";

    private static string Center(string value, int width)
    {
        if (value.Length >= width)
        {
            return value;
        }

        var padding = (width - value.Length) / 2;
        return new string(' ', padding) + value;
    }

    private static string PadRight(string value, int width) =>
        value.PadRight(Math.Max(value.Length, width));

    private static IReadOnlyCollection<string> Wrap(string? value, int width)
    {
        if (string.IsNullOrWhiteSpace(value))
        {
            return Array.Empty<string>();
        }

        var words = value.Split(' ', StringSplitOptions.RemoveEmptyEntries);
        var lines = new List<string>();
        var current = string.Empty;

        foreach (var word in words)
        {
            var candidate = string.IsNullOrEmpty(current) ? word : $"{current} {word}";
            if (candidate.Length <= width)
            {
                current = candidate;
                continue;
            }

            if (!string.IsNullOrEmpty(current))
            {
                lines.Add(current);
            }

            current = word;
        }

        if (!string.IsNullOrEmpty(current))
        {
            lines.Add(current);
        }

        return lines;
    }

    private static string TicketHeader(int width)
    {
        const string header = "CANT PRODUCTO P.UNIT IMPORTE";
        return header.Length <= width ? header : header[..width];
    }
}

File: backend-java/minimarket-print-worker-java/src/main/java/com/minimarket/printworker/service/TicketTextRenderer.java
package com.minimarket.printworker.service;

import com.minimarket.printworker.config.PrintingProperties;
import com.minimarket.printworker.dto.TicketPrintPayloadDto;
import org.springframework.stereotype.Component;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;

@Component
public class TicketTextRenderer {
    private static final String DATE_PATTERN = "dd/MM/yyyy HH:mm";

    public String build(TicketPrintPayloadDto payload, PrintingProperties properties) {
        var width = Math.max(32, properties.getLineWidth());

        // Los datos vienen del snapshot (igual que la vista previa). Si un ticket viejo no los trae,
        // se usan como respaldo las propiedades locales del worker.
        var businessName = fallback(payload.businessName(), properties.getBusinessName());
        var legalName = fallback(payload.legalName(), properties.getLegalName());
        var taxId = fallback(payload.taxId(), properties.getTaxId());
        var addressLine = fallback(payload.addressLine(), properties.getAddressLine());
        var phone = fallback(payload.phone(), properties.getPhone());
        var tagline = payload.tagline() != null ? payload.tagline().trim() : "";
        var documentTitle = fallback(payload.documentTitle(), "Ticket de venta");
        var customerLabel = fallback(payload.customerLabel(), properties.getCustomerLabel());
        var footerLine1 = fallback(payload.footerLine1(), properties.getFooterLine1());
        var footerLine2 = fallback(payload.footerLine2(), properties.getFooterLine2());

        // Totales del snapshot; si un ticket viejo no trae el desglose, se deriva del total (IGV 18% incluido).
        var total = payload.total();
        var subTotal = payload.subTotal() != null && payload.subTotal().signum() > 0
            ? payload.subTotal()
            : total.divide(BigDecimal.valueOf(1.18), 2, java.math.RoundingMode.HALF_UP);
        var igv = payload.igv() != null && payload.igv().signum() > 0
            ? payload.igv()
            : total.subtract(subTotal);

        var lines = new ArrayList<String>();
        lines.add(center(businessName, width));
        if (!tagline.isBlank()) {
            lines.add(center(tagline, width));
        }
        lines.add("-".repeat(width));
        lines.add(center(legalName, width));
        lines.add(center("RUC: " + taxId, width));
        lines.add(center(addressLine, width));
        lines.add(center("Telefono: " + phone, width));
        lines.add("-".repeat(width));
        lines.add(center(documentTitle.toUpperCase(), width));
        lines.add(center("#" + payload.saleId() + " - " + payload.saleDate().format(java.time.format.DateTimeFormatter.ofPattern(DATE_PATTERN)), width));
        lines.add("-".repeat(width));
        lines.add(metaRow("Fecha", payload.saleDate().format(java.time.format.DateTimeFormatter.ofPattern(DATE_PATTERN)), width));
        lines.add(metaRow("Cajero", payload.cashierName(), width));
        lines.add(metaRow("Cliente", customerLabel, width));
        lines.add(metaRow("Pago", payload.paymentMethod(), width));
        lines.add("-".repeat(width));
        lines.add(ticketHeader(width));

        for (var item : payload.items()) {
            lines.addAll(ticketItemLines(item, width));
            lines.add("-".repeat(width));
        }

        lines.add(row("Items", String.valueOf(payload.items().size()), width));
        var totalUnits = payload.items().stream().mapToInt(item -> item.quantity() == null ? 0 : item.quantity()).sum();
        lines.add(row("Unidades", String.valueOf(totalUnits), width));
        lines.add(row("Subtotal", formatMoney(subTotal), width));
        lines.add(row("IGV (18%)", formatMoney(igv), width));

        if (payload.notes() != null && !payload.notes().isBlank()) {
            lines.add("-".repeat(width));
            lines.add("Notas:");
            lines.addAll(wrap(payload.notes(), width));
        }

        lines.add("-".repeat(width));
        lines.add(row("Total", formatMoney(total), width));
        lines.add("");
        lines.add(center(footerLine1, width));
        lines.add(center(footerLine2, width));
        lines.add("");
        lines.add("");

        return String.join(System.lineSeparator(), lines);
    }

    private static String fallback(String value, String fallback) {
        return value == null || value.isBlank() ? fallback : value.trim();
    }

    private static String row(String label, String value, int width) {
        return padRight(label, Math.max(0, width - value.length())) + value;
    }

    private static String metaRow(String label, String value, int width) {
        return row(label, value, width);
    }

    private static String formatMoney(BigDecimal value) {
        return "S/ " + value.setScale(2, java.math.RoundingMode.HALF_UP);
    }

    private static String center(String value, int width) {
        if (value.length() >= width) {
            return value;
        }

        var padding = (width - value.length()) / 2;
        return " ".repeat(Math.max(0, padding)) + value;
    }

    private static String padRight(String value, int width) {
        return String.format("%1$-" + Math.max(value.length(), width) + "s", value);
    }

    private static List<String> wrap(String value, int width) {
        var lines = new ArrayList<String>();
        if (value == null || value.isBlank()) {
            return lines;
        }

        var words = value.split("\\s+");
        var current = new StringBuilder();
        for (var word : words) {
            var candidate = current.length() == 0 ? word : current + " " + word;
            if (candidate.length() <= width) {
                current = new StringBuilder(candidate);
                continue;
            }

            if (current.length() > 0) {
                lines.add(current.toString());
            }

            current = new StringBuilder(word);
        }

        if (current.length() > 0) {
            lines.add(current.toString());
        }

        return lines;
    }

    private static String ticketHeader(int width) {
        var layout = columnLayout(width);
        return padRight("CANT.", layout.qtyWidth())
            + " "
            + padRight("PRODUCTO", layout.productWidth())
            + " "
            + padLeft("P. UNIT", layout.unitWidth())
            + " "
            + padLeft("IMPORTE", layout.amountWidth());
    }

    private static List<String> ticketItemLines(com.minimarket.printworker.dto.TicketPrintItemDto item, int width) {
        var layout = columnLayout(width);
        var productLines = wrap(item.productName(), layout.productWidth());
        if (productLines.isEmpty()) {
            productLines = List.of("");
        }

        var lines = new ArrayList<String>();
        var firstLine = padRight(String.valueOf(item.quantity()), layout.qtyWidth())
            + " "
            + padRight(productLines.getFirst(), layout.productWidth())
            + " "
            + padLeft(formatMoney(item.unitPrice()), layout.unitWidth())
            + " "
            + padLeft(formatMoney(item.subtotal()), layout.amountWidth());
        lines.add(firstLine);

        for (var index = 1; index < productLines.size(); index++) {
            lines.add(padRight("", layout.qtyWidth())
                + " "
                + padRight(productLines.get(index), layout.productWidth()));
        }

        return lines;
    }

    private static ColumnLayout columnLayout(int width) {
        var qtyWidth = 5;
        var unitWidth = 10;
        var amountWidth = 10;
        var spaces = 3;
        var productWidth = Math.max(12, width - qtyWidth - unitWidth - amountWidth - spaces);
        return new ColumnLayout(qtyWidth, productWidth, unitWidth, amountWidth);
    }

    private static String padLeft(String value, int width) {
        return String.format("%" + Math.max(value.length(), width) + "s", value);
    }

    private record ColumnLayout(int qtyWidth, int productWidth, int unitWidth, int amountWidth) {
    }
}
